File: binaryTrees/42.cpp
#include "../common/tree.h"

using namespace Tree;

void solve(TreeNode<int>* root, vector<vector<int>>& answer, int cont, vector<int> aux,
           int k) {
    if (root == nullptr) return;

    cont += root->val;
    aux.push_back(root->val);

    if (root->left == nullptr && root->right == nullptr && cont == k) {
        answer.push_back(aux);
        return;
    }

    solve(root->left, answer, cont, aux, k);
    solve(root->right, answer, cont, aux, k);
}

vector<vector<int>> pathSum(TreeNode<int>* root, int targetSum) {
    int cont = 0;
    vector<vector<int>> answer;
    vector<int> temp;

    solve(root, answer, cont, temp, targetSum);

    return answer;
}

int main() {
    TreeNode<int>* root = createNode(5);
    root->left = createNode(4);
    root->right = createNode(8);

    root->left->left = createNode(11);

    root->left->left->left = createNode(7);
    root->left->left->right = createNode(2);

    root->right->left = createNode(13);
    root->right->right = createNode(4);

    root->right->right->left = createNode(5);
    root->right->right->right = createNode(1);

    auto res = pathSum(root, 22);

    for (const auto& v : res) {
        for (const auto& e : v) {
            cout << e << " ";
        }

        cout << "\n";
    }

    return 0;
}

File: binaryTrees/42.py
from typing import List, Optional
import copy


class TreeNode:
    def __init__(self, k: int):
        self.val: int = k
        self.left: TreeNode = None
        self.right: TreeNode = None


def pathSum(root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
    temp: List[int] = []
    ans: List[int] = []
    cont = 0

    def dfs(root, ans, targetSum, temp, cont):
        if root is None:
            return

        cont += root.val
        temp.append(root.val)

        if root.left is None and root.right is None and cont == targetSum:
            ans.append(copy.deepcopy(temp))
            temp.pop()
            return

        dfs(root.left, ans, targetSum, temp, cont)
        dfs(root.right, ans, targetSum, temp, cont)

        temp.pop()

    dfs(root, ans, targetSum, temp, cont)

    return ans


targetSum = 22
root = TreeNode(5)
root.left = TreeNode(4)
root.right = TreeNode(8)
root.left.left = TreeNode(11)
root.left.left.left = TreeNode(7)
root.left.left.right = TreeNode(2)

root.right.left = TreeNode(13)
root.right.right = TreeNode(4)
root.right.right.left = TreeNode(5)
root.right.right.right = TreeNode(1)

print(pathSum(root, 22))
